src/test/java/org/jenkinsci/plugins/workflow/cps/steps/ParallelStepTest.java

```diff
@@ -517,4 +528,193 @@ public void parallelLexicalScope() throws Exception {
+        assertEquals(
+                1,
+                comparator.compare(
+                        new FlowInterruptedException(
+                                Result.SUCCESS, new TimeoutStepExecution.ExceededTimeout()),
+                        new FlowInterruptedException(
+                                Result.UNSTABLE, new TimeoutStepExecution.ExceededTimeout())));
+        assertEquals(
+                -1,
+                comparator.compare(
+                        new FlowInterruptedException(
+                                Result.UNSTABLE, new TimeoutStepExecution.ExceededTimeout()),
+                        new FlowInterruptedException(
+                                Result.SUCCESS, new TimeoutStepExecution.ExceededTimeout())));
+        assertEquals(
+                1,
+                comparator.compare(
+                        new FlowInterruptedException(
+                                Result.UNSTABLE, new TimeoutStepExecution.ExceededTimeout()),
+                        new FlowInterruptedException(
+                                Result.FAILURE, new TimeoutStepExecution.ExceededTimeout())));
+        assertEquals(
+                -1,
+                comparator.compare(
+                        new FlowInterruptedException(
+                                Result.FAILURE, new TimeoutStepExecution.ExceededTimeout()),
+                        new FlowInterruptedException(
+                                Result.UNSTABLE, new TimeoutStepExecution.ExceededTimeout())));
+    }
+
+    @Issue("JENKINS-49073")
+    @Test
+    public void parallelPropagatesStatus() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                parallelPropagatesStatusImpl(
+                        Result.SUCCESS, Result.SUCCESS, Result.SUCCESS, Result.SUCCESS);
+                parallelPropagatesStatusImpl(
+                        Result.UNSTABLE, Result.SUCCESS, Result.SUCCESS, Result.UNSTABLE);
+                parallelPropagatesStatusImpl(
+                        Result.FAILURE, Result.SUCCESS, Result.SUCCESS, Result.FAILURE);
+                parallelPropagatesStatusImpl(
+                        Result.FAILURE, Result.SUCCESS, Result.UNSTABLE, Result.FAILURE);
+                parallelPropagatesStatusImpl(
+                        Result.FAILURE, Result.SUCCESS, Result.FAILURE, Result.UNSTABLE);
+                parallelPropagatesStatusImpl(
+                        Result.NOT_BUILT, Result.SUCCESS, Result.NOT_BUILT);
+                parallelPropagatesStatusImpl(
+                        Result.NOT_BUILT, Result.SUCCESS, Result.UNSTABLE, Result.NOT_BUILT);
+                parallelPropagatesStatusImpl(
+                        Result.NOT_BUILT, Result.SUCCESS, Result.UNSTABLE, Result.FAILURE, Result.NOT_BUILT);
+                parallelPropagatesStatusImpl(
+                        Result.NOT_BUILT, Result.NOT_BUILT, Result.NOT_BUILT);
```

Comment:
I'm not entirely satisfied with how the code is currently working for these test cases. For example, for three parallel builds whose results are `SUCCESS`, `UNSTABLE`, and `NOT_BUILT` we are currently propagating a status of `NOT_BUILT`. I don't consider this ideal; to me, `UNSTABLE` is "worse" than `NOT_BUILT`. Similarly, for four parallel builds whose results are `SUCCESS`, `UNSTABLE`, `FAILURE`, and `NOT_BUILT` we are currently propagating a status of `NOT_BUILT. I don't consider this ideal; to me, `FAILURE` is "worse" than `NOT_BUILT`. Finally, for two parallel builds whose status is `SUCCESS` and `NOT_BUILT` we are currently propagating a status of `NOT_BUILT`. This doesn't make sense to me. If all the child jobs had a status of `NOT_BUILT`, then I could understanding the parent job having a status of `NOT_BUILT`, but as long as a single child job has a status of `SUCCESS`, I would expect that to be propagated to the parent.

I will think some more about how to tweak the comparator algorithm to handle these cases.

Reply:
Another tricky question is how to handle `ABORTED`. Between `SUCCESS` and `ABORTED`, `ABORTED` is clearly worse. But what about between `FAILURE`/`UNSTABLE` and `ABORTED`? I would argue that `FAILURE`/`UNSTABLE` are "worse", because we _know_ the status of the failed/unstable job (whereas we _don't_ know the status of the aborted job, which never ran to completion). Between `NOT_BUILT` and `ABORTED` I think `ABORTED` is worse.

Reply:
Re: What result is worst, I would just use the ordering defined by `Result.ordinal` (used to implement `Result.isWorseThan` and friends). Whether that ordering makes the most sense could be debated, but that's the ordering that everything else in Jenkins uses, so I would use it here as well.

Might be interesting to run pipeline-model-definition's tests with this change to see if any of them start failing.

Reply:
> Re: What result is worst, I would just use the ordering defined by `Result.ordinal` (used to implement `Result.isWorseThan` and friends). Whether that ordering makes the most sense could be debated, but that's the ordering that everything else in Jenkins uses, so I would use it here as well.

My comparison logic is already using `Result#isWorseThan` and `Result#isBetterThan`, so if we're OK with the behavior demonstrated by the new tests (which I think is acceptable, but not ideal) then I can leave the comparison logic as-is.

> Might be interesting to run pipeline-model-definition's tests with this change to see if any of them start failing.

Sure. I opened jenkinsci/pipeline-model-definition-plugin#360 to cover this testing.

Reply:
> I opened [jenkinsci/pipeline-model-definition-plugin#360](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/360) to cover this testing.

These tests passed.

Reply:
To elaborate on my earlier comment, the values in `Result` have always seemed weird to me. `ABORTED` and `NOT_BUILT` feel like a different kind of thing entirely than `SUCCESS`, `FAILURE`, and `UNSTABLE`. Sometimes I think it would have made more sense for results to have 2 parts, a "completion" result, like , `NOT_BUILT`, `ABORTED`, `COMPLETED`, and a "goodness" result, like `SUCCESS`, `WARNING`, `ERROR`. Then you could have something like a Pipeline step that is `ABORTED` and `SUCCESS` in the case of a user cancelling the build, or `ABORTED` and `WARNING` in the case of something like a timeout, `NOT_BUILT` and `SUCCESS` could be used in cases like a Declarative `when` expression that is false, and here in the parallel step, the completion result would always be `COMPLETED`, and only the "goodness" result would be determined by the results of the branches. Using two values creates other problems though, and not all combinations make sense.

So all that to say, I agree that the behavior here with `NOT_BUILT` feels wrong, but I'm not sure if it makes sense to try to change it.

Reply:
I agree with your observations here. The awkwardness in the use cases I described above is exposing a deeper problem in how the `Result` API is defined in Jenkins core. When I get a chance I'll try to file a bug/RFE against Jenkins core with the results of this discussion. Perhaps one day that can be addressed, but I agree that it is quite orthogonal to this effort.